keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+resources objects.
+
+One common use case is [ConfigMaps][] and [Secrets][]. The users roll out
+ConfigMaps and Secrets by creating new ConfigMaps or Secrets and updating the
+references to those ConfigMaps or Secrets in containers ([#22368][]). The old
+ConfigMaps and Secrets, if not referenced by any container, are no longer
+needed. 
+
+Another common use case is completed [Jobs][]. Users keep creating Jobs to run
+some batch tasks, and those Jobs, once complete, are no longer needed in the
+system.
+
+Failing to clean up those resources could cause cluster instability (e.g. API
+server requests timeouts and health-check failures) and cost more compute
+resource usage than actually needed -- this is not a hypothetical problem but a
+real, long-standing Kubernetes usability issue.
+
+[ConfigMaps]: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/
+[Secrets]: https://kubernetes.io/docs/concepts/configuration/secret/
+[#22368]: https://github.com/kubernetes/kubernetes/issues/22368
+[Jobs]: https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/
+
+### Goals
+
+A successful mechanism must:
+
+* Be declarative. 
+* Be able to deal with custom (extension) resources and custom controllers. 
+* Allow users to opt-in and opt-out of this feature without breaking
+  functionality.
+* Can be safely enabled / rolled out and then disabled / rolled back.
+* Maintain backward compatibility. 
+* Have a low cognitive burden. Easy to set up and reason about what resources
+  should be garbage collected.
+* Be efficient enough that it wouldn’t introduce a new scalability bottleneck.
+
+### Non-Goals
+
+Other related features, such as the design of resource immutability,
+cluster-wide resource quota, or undo resource deletion, are orthogonal to
+garbage collection and their design won’t be discussed in this proposal.
+
+## Background
+
+### Garbage Collection in Kubernetes
+
+Currently in Kubernetes, there are 3 flavors of garbage collection: 
+ownerReference-based, capacity-based, and TTL-based.
+
+#### OwnerReference-based Garbage Collection  (Generic)
+
+[OwnerReference-based garbage collection][] supports cascading deletion for all
+Kubernetes resources. This is managed by a controller named generic garbage
+collector. The generic garbage collector deletes Kubernetes objects that once
+had an owner, but no longer have an owner. A Kubernetes object defines its owner
+by adding its owner to its own  `.metadata.ownerReferences` field. You can think
+of `.metadata.ownerReferences` as a list of pointers to a resource's parents
+(owners). 
+
+For example, when a Deployment owns a ReplicaSet, that ReplicaSet's
+`.metadata.ownerReferences` will include the Deployment. If the generic garbage
+collector finds that the Deployment doesn't exist in the system anymore, and if
+the Deployment is the ReplicaSet's only owner, it will delete the ReplicaSet. 
```

Comment:
what about the case when the selector is changed. Currently we have a case where because of selector changes, all of our olds Pods became orphaned. I have the following questions:-
- The DaemonSet's selector changed,which orphaned the old pods and DaemonSet created new pods. I saw that the ownerReferences of the old Pods doesnt exist any more.  My question is who removed those ownerReferences ?
- Does your proposal cover deleting such orphans whose parents are gone because of selector change ?
If we didnt remove the ownerReferences of these old Pods , but changed the uid of the owner since the selector of the owner change, we could achieve this. Thoughts @janetkuo ?

Reply:
> Currently we have a case where because of selector changes, all of our olds Pods became orphaned. 

That is exactly why we made selector immutable in apps/v1. Changing selector of a workload controller is dangerous, and this specific use case is not something we want to support.

Reply:
thanks @janetkuo  i am fine with this. But why when the selector changed, did the orphan'd pods loose their ownerReferences ?

Reply:
ControllerReferences (OwnerReferences with controller==true) of Pods are updated automatically by workload controllers, to make sure that no two controllers are fighting against each others controlling the same group of Pods (think of it as a lock). When a controller no longer controls a group of Pods (selector no longer matches), it removes itself from the Pods' ControllerReferences to release the lock. 

Label selectors are used for identifying relationships between controllers and controllee. When a controller's selector is changed to something that doesn't match its controllees' labels, it has the same effect as orphaning the controllees by changing their labels, as the controllers finds controllees through "label/selector matches". 

If you really want to change the selector of a controller, the recommended way is to just create a new one with the new selector / template labels, and delete the old one.

Reply:
Thanks @Janetkuo for helping understand. the intention is not to change the selectors . We changed it by accident but just be able to properly reason about its behavior . How does the controller know that a pod’s label no longer matches it’s selector . The controller can know which pods matches its label and rest everything else does not match its label . So does the controlller actually goes and looks at all pods that didn’t match its selector and see if it has an owner reference pointing to it and then Remove it self from those pods? Isn’t that too costly? Is the selector immutable in apps/v1 for all workload controllers ? Do you see any issues with GC always deleting a pod which doesn’t have any ownerreferneces or if it’s owner doesn’t exist . Why does t have to be triggered by owners deletion ? Sorry lot of questions :-)

Reply:
> So does the controlller actually goes and looks at all pods that didn’t match its selector and see if it has an owner reference pointing to it and then Remove it self from those pods? 

This is correct. 

> Isn’t that too costly? 

Most controllers use informer framework, so instead of listing all controllees (e.g. pods), the controller watches them and acts upon their changes. So it's not too costly. 

> Do you see any issues with GC always deleting a pod which doesn’t have any ownerreferneces or if it’s owner doesn’t exist. Why does t have to be triggered by owners deletion ?

GC doesn't delete things that don't have ownerRefs. It deletes things whose owners are all gone. Think of ownerRefs as a signal for cascading deletion: the owner wants its dependents to be gone with it; if there are more than one owners, the GC should only happen after all owners are deleted. Note that if an item is removed from a dependent's ownerRefs list, that item is no longer considered the owner of the dependent.